### ISSUE TITLE
Need a clustering algo where clusters have equal sizes

### ISSUE THREAD

**Issue by marcinjurek** [katzfuss-group/GPvecchia#29]
We need an algorithm to efficiently partition a domain into clusters that are contiguous in some intuitive sense (when you see it, you know it's right) but at the same time have an equal number of points. The obvious starting point, k-means, doesn't ensure that. In fact, clusters can have very different sizes. 

The most promising approach seemed to be some way of modifying k-means. For example, some people on Stack Overflow suggest to first run k-means, then order the (point, cluster, distance) tuples by distance and go through this list to add points to clusters unless clusters have a required size.

I also tried a slightly modified version of this approach where we don't go through the list, but find the smallest cluster, and add the point **to that cluster** that is the nearest. 

Finally, I briefly talked to Dr. Sang about this and she suggested to modify k-means so that while we compute the clusters, we stop adding points to the cluster when we reach its desired size.

All of these approaches suffer from the same problem: while the resulting clusters do have the same number of points, they are often drastically non-contiguous. Find enclosed my R code that implements all three versions (had to save it as txt or github won't let me upload it). Maybe one of you has some idea? 

[test-kmeans.txt](https://github.com/katzfuss-group/GPvecchia/files/2621197/test-kmeans.txt)

**Comment by katzfuss** [comment]
How about this: Let's say we want J groups of roughly equal size. We choose the smallest k such that 2^k >= J. Then, we do the splitting algorithm we use for J=2 (split along one dimension using the median) k times, ending up with 2^k groups of roughly equal size. While the size of each group then is often a bit smaller than what we wanted (depending on the exact value of J), it should give us spatially contiguous groups that respect the desired maximum group size.

**Comment by marcinjurek** [comment]
Ok, I see the idea. It seems to me, however, that this would give us partitions for J that is a power of 2. For example, if J=25, then I'm not sure how that would work.

But do we really need a fully general J? It seems to me, that we do only in FSA. So maybe let's assume that J is indeed always a power of 2 and when the user requests something else, we warn him that we used the nearest power of 2 (which is also >J) instead?

**Comment by marcinjurek** [comment]
accidentally closed it:/

**Comment by katzfuss** [comment]
Yes, the number of partitions would be a power of 2. So for J=25, we would use 2^5=32 partitions, which would give slightly smaller cluster sizes than requested. You can give out a warning in that case. Please go ahead and implement, I would like this MRA-Vecchia to be completed soon.
